In [17]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import sklearn as sk


In [29]:
data = pd.read_csv("kc_house_data.csv",dtype = {'bathrooms':float, 'waterfront':int, 'sqft_above':int, 'sqft_living15':float, 'grade':int, 'yr_renovated':int, 'price':float, 'bedrooms':float, 'zipcode':str, 'long':float, 'sqft_lot15':float, 'sqft_living':float, 'floors':str, 'condition':int, 'lat':float, 'date':str, 'sqft_basement':int, 'yr_built':int, 'id':str, 'sqft_lot':int, 'view':int})
data_train = pd.read_csv("kc_house_train_data.csv",dtype = {'bathrooms':float, 'waterfront':int, 'sqft_above':int, 'sqft_living15':float, 'grade':int, 'yr_renovated':int, 'price':float, 'bedrooms':float, 'zipcode':str, 'long':float, 'sqft_lot15':float, 'sqft_living':float, 'floors':str, 'condition':int, 'lat':float, 'date':str, 'sqft_basement':int, 'yr_built':int, 'id':str, 'sqft_lot':int, 'view':int})
data_test = pd.read_csv("kc_house_test_data.csv",dtype = {'bathrooms':float, 'waterfront':int, 'sqft_above':int, 'sqft_living15':float, 'grade':int, 'yr_renovated':int, 'price':float, 'bedrooms':float, 'zipcode':str, 'long':float, 'sqft_lot15':float, 'sqft_living':float, 'floors':str, 'condition':int, 'lat':float, 'date':str, 'sqft_basement':int, 'yr_built':int, 'id':str, 'sqft_lot':int, 'view':int})

In [30]:
data.head(10)

id             date      price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000   221900.0       3.0       1.00       1180.0   
1  6414100192  20141209T000000   538000.0       3.0       2.25       2570.0   
2  5631500400  20150225T000000   180000.0       2.0       1.00        770.0   
3  2487200875  20141209T000000   604000.0       4.0       3.00       1960.0   
4  1954400510  20150218T000000   510000.0       3.0       2.00       1680.0   
5  7237550310  20140512T000000  1225000.0       4.0       4.50       5420.0   
6  1321400060  20140627T000000   257500.0       3.0       2.25       1715.0   
7  2008000270  20150115T000000   291850.0       3.0       1.50       1060.0   
8  2414600126  20150415T000000   229500.0       3.0       1.00       1780.0   
9  3793500160  20150312T000000   323000.0       3.0       2.50       1890.0   

   sqft_lot floors  waterfront  view     ...      grade  sqft_above  \
0      5650      1           0     0     ...          7        1180   
1      7242      2           0     0     ...          7        2170   
2     10000      1           0     0     ...          6         770   
3      5000      1           0     0     ...          7        1050   
4      8080      1           0     0     ...          8        1680   
5    101930      1           0     0     ...         11        3890   
6      6819      2           0     0     ...          7        1715   
7      9711      1           0     0     ...          7        1060   
8      7470      1           0     0     ...          7        1050   
9      6560      2           0     0     ...          7        1890   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1955             0    98178  47.5112 -122.257   
1            400      1951          1991    98125  47.7210 -122.319   
2              0      1933             0    98028  47.7379 -122.233   
3            910      1965             0    98136  47.5208 -122.393   
4              0      1987             0    98074  47.6168 -122.045   
5           1530      2001             0    98053  47.6561 -122.005   
6              0      1995             0    98003  47.3097 -122.327   
7              0      1963             0    98198  47.4095 -122.315   
8            730      1960             0    98146  47.5123 -122.337   
9              0      2003             0    98038  47.3684 -122.031   

   sqft_living15  sqft_lot15  
0         1340.0      5650.0  
1         1690.0      7639.0  
2         2720.0      8062.0  
3         1360.0      5000.0  
4         1800.0      7503.0  
5         4760.0    101930.0  
6         2238.0      6819.0  
7         1650.0      9711.0  
8         1780.0      8113.0  
9         2390.0      7570.0  

[10 rows x 21 columns]

In [62]:
data['bedroom_squared'] = data['bedrooms']**2
data['bed_bath_rooms'] = data['bedrooms']*data['bathrooms']
data['log_sqft_living'] = data['sqft_living'].apply(np.math.log)
data['lat_plus_long'] = data['lat'] + data['long']

data_train['bedroom_squared'] = data['bedrooms']**2
data_train['bed_bath_rooms'] = data['bedrooms']*data['bathrooms']
data_train['log_sqft_living'] = data['sqft_living'].apply(np.math.log)
data_train['lat_plus_long'] = data['lat'] + data['long']

data_test['bedroom_squared'] = data['bedrooms']**2
data_test['bed_bath_rooms'] = data['bedrooms']*data['bathrooms']
data_test['log_sqft_living'] = data['sqft_living'].apply(np.math.log)
data_test['lat_plus_long'] = data['lat'] + data['long']


# Added new input features to the data set 
* bedrooms squared
* bed_bath_rooms
* log_sqft_living
* lat_plus_long

In [63]:
print "Mean of bedrooms_squared is : ",data_test['bedroom_squared'].mean()
print "Mean of bed_bath_rooms is : ",data_test['bed_bath_rooms'].mean()
print "Mean of log_sqft_living is :",data_test['log_sqft_living'].mean()
print "Mean of lat_plus_long is :",data_test['lat_plus_long'].mean()

Mean of bedrooms_squared is :  12.1875147789
Mean of bed_bath_rooms is :  7.32826909435
Mean of log_sqft_living is : 7.54004185613
Mean of lat_plus_long is : -74.6559050603


# Mean of bedrooms_squared is :  12.1875147789
# Mean of bed_bath_rooms is :  7.32826909435
# Mean of log_sqft_living is : 7.54004185613
# Mean of lat_plus_long is : -74.6559050603



In [64]:
model_1 = sk.linear_model.LinearRegression()
model_1.fit(data_train[['sqft_living','bedrooms','bathrooms','lat','long']],data_train['price'])
model_2 = sk.linear_model.LinearRegression()
model_2.fit(data_train[['sqft_living','bedrooms','bathrooms','lat','long','bed_bath_rooms']],data_train['price'])
model_3 = sk.linear_model.LinearRegression()
model_3.fit(data_train[['sqft_living','bedrooms','bathrooms','lat','long','bed_bath_rooms','bedroom_squared',
                       'log_sqft_living','lat_plus_long']],data_train['price'])

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

In [65]:
print "Model 1 coefficients : ",model_1.intercept_,model_1.coef_
print "Model 2 coefficients : ",model_2.intercept_,model_2.coef_
print "Model 2 coefficients : ",model_3.intercept_,model_3.coef_


Model 1 coefficients :  -69075726.7926 [  3.12258646e+02  -5.95865332e+04   1.57067421e+04   6.58619264e+05
  -3.09374351e+05]
Model 2 coefficients :  -69076568.5239 [  3.12306473e+02  -5.95872323e+04   1.56844899e+04   6.58597198e+05
  -3.09351344e+05   6.36272345e+02]
Model 2 coefficients :  -69640349.6593 [  3.12278591e+02  -5.95775064e+04   1.57113977e+04   6.58576991e+05
  -3.09375795e+05   1.17036847e+03  -1.49224586e+02  -3.38252912e+03
  -7.83794967e+03]


In [68]:
output_1 = model_1.predict(data_test[['sqft_living','bedrooms','bathrooms','lat','long']])
output_2 = model_2.predict(data_test[['sqft_living','bedrooms','bathrooms','lat','long','bed_bath_rooms']])
output_3 = model_3.predict(data_test[['sqft_living','bedrooms','bathrooms','lat','long','bed_bath_rooms','bedroom_squared',
                       'log_sqft_living','lat_plus_long']])

output_1_train = model_1.predict(data_train[['sqft_living','bedrooms','bathrooms','lat','long']])
output_2_train = model_2.predict(data_train[['sqft_living','bedrooms','bathrooms','lat','long','bed_bath_rooms']])
output_3_train = model_3.predict(data_train[['sqft_living','bedrooms','bathrooms','lat','long','bed_bath_rooms','bedroom_squared',
                       'log_sqft_living','lat_plus_long']])


In [69]:
print "RSS 1 : ",((data_test['price'] - output_1)**2).sum()
print "RSS 2 : ",((data_test['price'] - output_2)**2).sum()
print "RSS 3 : ",((data_test['price'] - output_3)**2).sum()

RSS 1 :  2.25500469795e+14
RSS 2 :  2.25661972813e+14
RSS 3 :  2.25684652569e+14


# MODEL 1 HAS LOWEST RSS ON TEST SET

In [70]:
print "RSS 1 : ",((data_train['price'] - output_1_train)**2).sum()
print "RSS 2 : ",((data_train['price'] - output_2_train)**2).sum()
print "RSS 3 : ",((data_train['price'] - output_3_train)**2).sum()

RSS 1 :  9.6787996305e+14
RSS 2 :  9.67757281242e+14
RSS 3 :  9.67680784989e+14
